lib/rules/length-zero-no-unit/index.js

```diff
@@ -2,201 +2,162 @@
-
-				// Only pay attention if the value parses to 0
-				// and units with lengths
-				if (
-					Number.parseFloat(valueWithZero) !== 0 ||
-					!keywordSets.lengthUnits.has(parsedValue.unit.toLowerCase())
-				) {
-					return;
-				}
-
-				if (context.fix) {
-					fixPositions.unshift({
-						startIndex: valueWithZeroStart,
-						length: valueWithZeroEnd - valueWithZeroStart,
-					});
-
-					return;
-				}
-
-				report({
-					message: messages.rejected,
-					node,
-					index: valueWithZeroEnd - parsedValue.unit.length,
-					result,
-					ruleName,
-				});
+			const parsedValue = valueParser(value);
+
+			parsedValue.walk((valueNode) => {
+				return check(node, getIndex(node), valueNode);
 			});
 
-			if (fixPositions.length) {
-				fixPositions.forEach((fixPosition) => {
-					if (node.type === 'atrule') {
-						// Use `-1` for `@` character before each at rule
-						const realIndex =
-							fixPosition.startIndex - node.name.length - node.raws.afterName.length - 1;
+			if (needsFix) {
+				set(node, parsedValue.toString());
+			}
+		}
+
+		function checkDecl(node, value, set, getIndex) {
+			needsFix = false;
 
-						node.params = replaceZero(node.params, realIndex, fixPosition.length);
-					} else {
-						const realIndex = fixPosition.startIndex - node.prop.length - node.raws.between.length;
+			const { prop } = node;
 
-						node.value = replaceZero(node.value, realIndex, fixPosition.length);
-					}
-				});
+			if (isLineHeight(prop)) return;
+
+			if (optionsMatches(secondary, 'ignore', 'custom-properties') && isCustomProperty(prop))
+				return;
+
+			const parsedValue = valueParser(value);
+
+			parsedValue.walk((valueNode, valueNodeIndex, valueNodes) => {
+				if (isLineHeightValue(node, valueNodes, valueNodeIndex)) return;
+
+				return check(node, getIndex(node), valueNode);
+			});
+
+			if (needsFix) {
+				set(node, parsedValue.toString());
 			}
 		}
+
+		root.walkAtRules((atRule) =>
+			checkAtRule(atRule, getAtRuleParams(atRule), setAtRuleParams, atRuleParamIndex),
+		);
+		root.walkDecls((decl) =>
+			checkDecl(decl, getDeclarationValue(decl), setDeclarationValue, declarationValueIndex),
+		);
 	};
 }
 
-function replaceZero(input, startIndex, length) {
-	const stringStart = input.slice(0, startIndex);
-	const stringEnd = input.slice(startIndex + length);
+function isLineHeightValue({ prop }, nodes, index) {
+	return (
+		prop.toLowerCase() === 'font' &&
+		index > 0 &&
+		nodes[index - 1].type === 'div' &&
+		nodes[index - 1].value === '/'
+	);
+}
+
+function isLineHeight(prop) {
+	return prop.toLowerCase() === 'line-height';
+}
+
+function isWord({ type }) {
+	return type !== 'word';
+}
+
+function isLength(unit) {
+	return keywordSets.lengthUnits.has(unit.toLowerCase());
+}
+
+function isFraction(unit) {
+	return unit.toLowerCase() === 'fr';
+}
 
-	return `${stringStart}0${stringEnd}`;
+function isZero(number) {
+	return Number.parseFloat(number) !== 0;
```

Comment:
[suggestion] The name `isZero` mismatches with the function body. How about fixing it as below?

```suggestion
	return Number.parseFloat(number) === 0;
```

And,

```js
if (!isZero(number)) return;
```